### ISSUE TITLE
faster updates full/pairmatch calls that change only the tol argument

### ISSUE THREAD

**Issue by benthestatistician** [markmfredrickson/optmatch#76]

**Comment by benthestatistician** [comment]
Adam Rauh (@adamrauh) is working on this problem now (summer 2017), on a separate fork of the hinting branch.  

As I see it, the next things we need are:

1. An object type in which to store "reduced costs" from a solved matching problem, but on a scale corresponding to the original, double-precision matching distance, rather than on the integer scale of "arc costs" (that `SubDivStrat` transform matching distances into). 
2.  An enhancement or replacement for `SubDivStrat` that (i) returns reduced costs, having pulled them out of the solved matching problem and back-transformed them to a scale matching the double-precision distances; (ii) accepts them, perhaps optionally,  and transforms them back down to integers in parallel with the transformation of distances before passing them down to `.fullmatch`. 

If I'm not mistaken, there'll be a reduced cost for each member of the treatment or control group found to be eligible for matching, plus a few more for bookkeeping nodes.  This suggests (a) storing these reduced costs as vectors aligned with the matching results, with an NA for each unit that was ineligible; (b) separately storing a short vector of reduced costs for bookkeeping nodes.   This might for example be a single named numeric vector, with designated names for the bookkeeping nodes. (One could use a weird name that doesn't commonly occur in the wild.)

**Comment by benthestatistician** [comment]
Correction to prev comment: reduced costs associate w/ arcs, not nodes.  See [definition](http://web.mit.edu/dimitrib/www/netbook_Full_Book.pdf#page=237) in Bertsekas's book, along with [explanation of how arcs' reduced costs classify the arcs as inactive, balanced or active](http://web.mit.edu/dimitrib/www/netbook_Full_Book.pdf#page=227) depending as they are inactive, balanced or active. 

Thus the objects we use for storing matching distances, InfinitySparseMatrix-es or ISMs, would be most directly applicable as storage vessels for these reduced costs, not vectors of length similar to that of the matching vector.  Storing all this information will sometimes be a bit costly in terms of memory, but in many problems those costs won't be too difficult to bear. 

Another approach that would be more sparing in terms of memory would be to reconstruct the price vector from the array of reduced costs, as follows.

1.  Set the price of the "end" node to an arbitrary, fixed, problem independent value, say 0.
2.  For each arc (j, e) incident to the end node figure difference $p_j - p_e$ by comparing reduced cost of arc with original arc cost.
3.  For each arc (j, e) incident to the end node, now figure price $p_j$. So that we now have prices for each node corresponding to a control group unit. 
4. For each node i corresponding to the treatment group, consider the first (by whatever ordering is convenient) arc connecting it to some member of the control group j.  Reconstruct $p_i - p_j$ bo comparing reduced cost to original cost of arc, and reconstruct p_i.
5.  This identifies all node prices except that of the sink node. But it can be now be reconstructed in similar fashion, by examination of reduced and original prices of any arc connecting it to some treatment group node.

**Comment by benthestatistician** [comment]
Update: We decided today that the next milestone to work toward is a version of SubDivStrat that optionally accepts one or more arguments recording reduced costs in double precision, and always returns double precision reduced costs.  In both cases, this information might be carried in one or several InfinitySparseMatrix'es (ISMs).  


(By using several ISMs, there could be one of precisely the dimensions of SubDivStrat's distance arguments, with one or more others to store the reduced costs associated with the outer arcs in [this picture](http://dept.stat.lsa.umich.edu/~bbh/hansenKlopfer2006.pdf#page=13); alternatively, by packing all this info into one ISM you get the advantage of only keeping track of a single data structure.  I somewhat favor the first of these two options.)  

There are various intermediate steps.  One that @adamrauh and I discussed was to modify the src/relax4s.f code so as to ensure that provided reduced costs are not ignored, with corresponding modification to R/fmatch.R to see that the default setting of reduced costs happens on the R side. A PS to that conversation:   Adam, once this is done please also update the inline comments at [line 212 of relax4s.f](https://github.com/markmfredrickson/optmatch/blob/hinting/src/relax4s.f#L212), i.e. immediately 
following 

    C   INITIALIZE DUAL PRICES
   
When the default-setting moves to the R side, the content of the comment should be moved with it.  New comments added to the fortran side can be written in lowercase, to distinguish them from the comments that came with the original code.

**Comment by benthestatistician** [comment]
[hinting ae9b6b0] does three things: pulls updates on master over the last 18 months or so into the hinting branch; adjusts`fmatch` to set the ["stablity increment"](http://dept.stat.lsa.umich.edu/~bbh/hansenKlopfer2006.pdf#page=15) explicitly, and implements a minimal change to get fmatch to pass up the reduced costs. Passes `devtools::test()` but not `devtools::check()`, but I suspect the latter stems from an issue with my system, not the optmatch code. So @adamrauh here are some suggested next steps: 

- [ ] Update your optmatch fork to pull in these updates to the hinting branch, despite the fact that there may be a bit of overlap with work you've already committed (and hence git merge conflicts to [resolve](https://help.github.com/articles/resolving-a-merge-conflict-using-the-command-line/)).
- [ ] Move stability incrementation: [master d6950f8] flags where the increment is now being set, inside of fmatch. But it seems to me that this operation groups naturally with conversion of double-precision distances to integer, so it should be occur in `SubDivStrat`, not `fmatch`, sometime after the distances are multiplied up and rounded but before they get passed down to fmatch.
- [ ] Reconstruct node prices within fmatch: I now think that fmatch should pass up node prices, not reduced costs -- `nc + nt + 2` integers, not `narcs` of them. (After thinking about it a little, extracting node prices from the reduced costs doesn't seem hard; I couldn't think of advantages of retaining reduced costs rather than node prices; and the reduced costs vector doesn't match the dimensions of `distance` anyway.)  
- [ ] Rejigger `fmatch`'s outputs, and corresponding expectations of functions that call it, to permit it also to return a node prices object that does not matching the dimensions of `distance`. 

A few comments ago I sketched the reconstruction of node prices from arc costs; here are some amendments to that comment:

- Arcs connecting treatment or control group nodes to sink or overflow nodes all have cost precisely 0. This means that the reduced costs of these arcs will precisely equal the differences of prices between the two nodes being connected.
- Since the network has an arc connecting each treatment or control group node to Overflow, aka "end", each of these nodes' prices _should_ equal the opposite of the reduced costs associated with those arcs.
- (I say "should" because I'm not sure how the relax solver treats arcs with upper capacity 0, which fmatch in some cases creates.  In essence these arcs aren't there.  Perhaps it returns NAs or nonsense values in these positions. We should find this out, documenting any assumptions we move forward on the basis of in the form of unit tests added to tests/testthat/test.fmatch.R.)

**Comment by adamrauh** [comment]
Working on implementing the process of constructing node prices using reduced and original costs as a separate function, which I will call within fmatch. I believe we determined that node prices would eventually be an optional parameter to subDivStrat (as opposed to reduced costs) which would then be converted and manipulated as necessary, before being passed to fmatch, again via a new optional argument.

How does one handle finding prices for nodes that are connected to multiple other nodes? For example, if there is an arc between node 1 and node 3, and node 2 and node 3, each with a different cost and reduced cost, wouldn't we get a different price for node 3 depending on which problem we solve?

**Comment by benthestatistician** [comment]
@adamrauh the reduced costs will work out in such a way that you get the same answer either way.  If not, there's a mistake somewhere.  (If you see an issue like this, check your work, and if it persists then do share on this thread the smallest self-contained example of it that you can find.)

**Comment by benthestatistician** [comment]
An update, drawn from email from @adamrauh:  

> Some progress:

> 1) Calculating the node prices is just a tad more complicated (but not much more) than a one or two line operation…there’s a little bit of extra logic involved. Not hard, but I do want to make sure it works as intended on a few more examples before I call this set.
> 
> 2) Provided 1) works as intended, fmatch will pass up a matrix with the following headers:
> 
>           control treated distance solution reduced.cost

All of this sounds good and in accord with what I think should be happening.  However, as to next steps after this, I see some issues both in what Adam wrote later in the same email and what I wrote  a few months back in the comment thread above. Accordingly I'll cross-out two remaining checklist items I see up there that now longer reflect my understanding of next steps, and I'll lay out some options here for those next steps. 

In terms of changes to user-visible objects, the goal we're now working toward,  as discussed offline with Adam and with @josherrickson last week, is to return objects of S3 class "`optmatch`" possessing, in addition to the attributes such objects now possess:

1.  An attribute encoding relevant information about node prices that was returned by the solver, indirectly by its returning of "reduced costs."  In addition,
2. It should be info about node prices, not reduced costs, node prices requiring about $n_t + n_c$ slots to store whereas reduced costs consume roughly $n_t * n_c$. 
3. The node prices should be expressed on a scale compatible with the double-precision distance that was originally passed down the `fullmatch` call stack, not the integer arc costs that would have been passed to and returned by the Fortran solver at the bottom of that call stack.
4. The information should be encoded in a single attribute, not multiple attributes, so that functions looking for the info can readily obtain yes-no answers about whether it's present. 

In order to get us to this point, the next proximate goal, after completing the changes that Adam details, needs to be to bring about (2) and (3). Based on Adam's report I think he's well on his way to (2). (2) should be completed at the `SubDivStrat` level of the call stack.  (This is consistent with the plan of Adam's detailed in the quoted email above regarding what to pass up from `fmatch`, although there are parts of what Adam wrote that suggest this translation is or will be done within `fmatch` itself.)  

Separately, (3) calls for a back-transformation of the scale of node prices, or arc costs.  This could be done before translating arc costs into node prices, or after.  Because it's wrapped up with the translation between "`tolerance`" and "`reso`" thats contained inside of `SubDivStrat`, inside of `SubDivStrat` is one natural place to do it.    One option is to do it there. 

However, computations related to tolerance and resolution are currently distributed between `SubDivStrat` and the next level up of the call stack, `fullmatch.matrix`; this is undesirable, complicating issues such as for instance #85, and ought really to be fixed.  So an approach to (3) that I somewhat prefer is to alter the `SubDivStrat` function signature so that (i) it expects a distspec argument containing integer, not double precision, distances, and (ii)  its `tolerance` argument goes away.  Instead, (iii) we create new helper functions, to be called inside of `fullmatch.matrix`, to: forward-transform the list of double-precision distances "`problems`" into a  list of integer distances, with scaling determined by the value of `fullmatch.matrix`'s `tol` argument; and later to back-transform  `SubDivStrat`'s  integer "`err`"s, and integer node prices that the updated `SubDivStrat` will return, to a double-precision scale matching that of the arguments that had been provided to `fullmatch.matrix`.

**Comment by adamrauh** [comment]
More notes on this topic outlining the details of next steps as I currently understand them, based on another conversation. Also to serve as a reference for myself:

- current process of extracting node prices should be implemented in subDivStrat, as opposed to being done in fmatch and passed back up. There may be some variables that make this easier in fmatch, but I will check this out.
- tolerance argument acts as a budget for how much "difference" a user is willing to tolerate between the objective value in the given solution that is generated and the best theoretically possible solution.

In cases where we have one tolerance value, but the problem ultimately gets divided up into multiple subclasses/subproblems, the program needs to figure out how much of the "tolerance budget" to distribute to each problem -- This occurs, at least in part, in fullmatch.matrix. Loosely speaking, these individual subproblems are then be "passed" to subDivStrat. Currently, subDivStrat will take the allowed tolerance as an argument and turn this into an integer by means of multiplication and rounding. It sounds like it may be multiple numbers that are used in this process of multiplication, but that it could be reduced to a single number. (This is something that needs to be verified, I think.) In theory, this process of multiplying and rounding could take place a level up in fullmatch.matrix, as opposed to subDivStrat. There should then be some "decoding" to convert the errors and any other information passed back from the solver back into double precision.

**Comment by adamrauh** [comment]
Regarding some of what is detailed above (specifically [here](https://github.com/markmfredrickson/optmatch/issues/76#issuecomment-335844027)), is there any reason that one couldn't make the assumption that both the "end" and "sink" nodes have a price of zero? The links between the "regular" nodes and the "end" and "sink" nodes don't get passed back up to SubDivStrat. It appears that all the control nodes have a link to the "sink" node. Since the controls are linked to both the "end" and "sink" nodes, the only way for the node prices to be equal would be for these two terminal nodes to be equal in price, I think (with zero just being the most logical choice). If this is the case, then my current implementation should work fine, which just boils down to: 

- multiplying the reduced costs found by the solver and passed up from fmatch by -1 and storing the results in a vector

- appending two zeroes to the end of the previous vector

- store this vector as an attribute of the object that is returned from fmatch

If not, then I believe either some information about the end/sink nodes will need to be passed up from fmatch to subDivStrat or simply do the calculations about node prices in fmatch and pass all that data back up

**Comment by benthestatistician** [comment]
On the issue54-hinting branch we can now accept a previous run as a hint.   

Whether/how much speed benefit this gives remains to be explored in detail, but I've observed no indication of it being the game-changer I had hoped for.   Perhaps relevant: we pass the dual solution (as "reduced costs") but not the matching primal solution (flow), as it would require a heavier touch to the Fortran to bring in the flow.  Perhaps that would give a better speed benefit?  

Musings aside, closing the issue, as that feels like a new issue.  (One that might be natural to raise in tandem w/ tamping down on memory use on the Fortran side - cf [Stop Copying Me](https://github.com/markmfredrickson/optmatchExperimental/projects/2) experimental project.)

**Comment by benthestatistician** [comment]
One reason, perhaps the only reason, that we didn't see speedups from hinting is that the problems we gave the solver in testing weren't challenging enough to the solver for hinting to make a difference. 
 See 5dcd5ba, which adds a test putting the solver through its paces.   In this test I'm seeing reliably and practically significant advantages from starting with a hint. 

That's with integer distances and hints, and a fully correct hint. A next step is to verify the same for non-integer distances.  It remains a possibility that the conversions between integer and double-precision distances and node prices aren't entirely in harmony with one another yet. (For example, `ceiling()` still makes sense to me as it's used for conversion of distances, but I wonder about `ceiling(node_info$price / epsilon)` for conversion of node prices to integer.  Why not `round(node_info$price / epsilon)`? [In `doubleSolve()`, in R/solve_reg_fm_prob.R.])

Once this is worked out, it will be interesting to see what benefits we get from ballpark correct hints.  And how much worse we can do by passing hints that are totally incorrect.

**Comment by benthestatistician** [comment]
This is now available in master, but only if you know it's there and how to get it. To do: 


- [ ]  daylight arguments to `fullmatch()`/`pairmatch()` that would be used to pass the hints. 
- [ ] clarify wherever appropriate that this requires the relax solver
- [ ] (Do we have checks that the hint MCFSolutions has the same subproblems structure as the current problem?)
- [ ] add tests of user-level functionality
- [ ] add examples

Question: Do we need to dissuade users from changing fullmatch arguments other than the tolerance? Or is it just as well to let them change whatever, as long as we can retrieve a node price for each of the relevant arguments?

**Comment by adamrauh** [comment]
@benthestatistician I wrote up some more detailed comments [here](https://github.com/adamrauh/optmatch/issues/2#issuecomment-2588454926) relevant to this issue (and some other issues). But I think we can close out this issue soon.  In brief, the remaining bits to be dealt with:

1) Where to put the relevant examples? I put some in MCFSolutions.Rmd, but I'm not sure if that's the best place.
2) Are these changes still only relevant to RELAX?
3) I'm not sure re: your last question. The code runs if other arguments are changed, but I don't know if that is substantively good.

**Comment by benthestatistician** [comment]
For my answers to 1-3 immediately above, pls see [this comment](https://github.com/adamrauh/optmatch/issues/2#issuecomment-2591566440).  Thanks much for that nice summary Adam!
